### PR TITLE
PDS: Add restore lib and test basic restore functionality.

### PR DESCRIPTION
**Pull request by gauraa2**
Summary: Add restore lib and test basic restore functionality.
Logs - https://aetos.pwx.purestorage.com/resultSet/testSetID/273020/testCaseID/1495367
Latest log - https://aetos.pwx.purestorage.com/resultSet/testSetID/273199

### CODE REVIEW

File: tests/pds/restore_test.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+				})
+			}
+		})
+	})
+	JustAfterEach(func() {
+		defer EndTorpedoTest()
+	})
+})
+
+func cleanupDeployment(dsInstances []*pds.ModelsDeployment) {
```
Comment (Madanagopal19) [inline tests/pds/restore_test.go:101]: move this method to common.go

**Review by Madanagopal19 — CHANGES_REQUESTED** [review]

**Review by Madanagopal19 — APPROVED** [review]
LGTM

**Review by aravindranath14 — APPROVED** [review]
LGTM

File: drivers/pds/pdsrestore/helper.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	}
+	return string(result)
+}
+
+func generateRandomName(prefix string) string {
+	nameSuffix := AlphaNumericString(NameSuffixLength)
+	return fmt.Sprintf("%s-systest-%s", prefix, nameSuffix)
+}
+
+func compareMaps(map1, map2 map[string]interface{}) bool {
```
Comment (lsrinivas-pure) [inline drivers/pds/pdsrestore/helper.go:36]: you can use reflect to compare maps
ex: reflect.DeepEqual(a, b)

File: drivers/pds/pdsrestore/restore_utils.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+func (restoreClient *RestoreClient) TriggerAndValidateRestore(backupJobId string, namespace string, bkpDsEntity DSEntity, isRestoreInSameNS, validate bool) (*pds.ModelsRestore, error) {
+	var (
+		bkpJob                 *pds.ModelsBackupJob
+		nsName, pdsNamespaceId string
+	)
+	k8sClusterId, err := restoreClient.RestoreTargetCluster.GetClusterID()
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch the cluster Id")
```
Comment (lsrinivas-pure) [inline drivers/pds/pdsrestore/restore_utils.go:42]: add err also in the log

File: drivers/pds/pdsrestore/restore_utils.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+	}
+	pdsRestoreTargetClusterId, err := restoreClient.RestoreTargetCluster.GetDeploymentTargetID(k8sClusterId, restoreClient.TenantId)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch the cluster details from the control plane")
```
Comment (lsrinivas-pure) [inline drivers/pds/pdsrestore/restore_utils.go:46]: same as above

File: drivers/pds/pdsrestore/restore_utils.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+		bkpJob, err = restoreClient.Components.BackupJob.GetBackupJob(backupJobId)
+		if err != nil {
+			return nil, err
+		}
+		nsName, pdsNamespaceId = namespace, bkpJob.GetNamespaceId()
+	}
+	log.Infof("backup Job - %v,Restore Target Cluster Id - %v, NamespaceId - %v", backupJobId, pdsRestoreTargetClusterId, pdsNamespaceId)
+	restoredModel, err := restoreClient.Components.Restore.RestoreToNewDeployment(backupJobId, "autom", pdsRestoreTargetClusterId, pdsNamespaceId)
+	if err != nil {
+		return nil, fmt.Errorf("failed during restore")
```
Comment (lsrinivas-pure) [inline drivers/pds/pdsrestore/restore_utils.go:63]: same here

File: drivers/pds/pdsrestore/restore_utils.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	}
+	err = wait.Poll(restoreTimeInterval, restoreTimeOut, func() (bool, error) {
+		restore, err := restoreClient.Components.Restore.GetRestore(restoredModel.GetId())
+		state := restore.GetStatus()
+		if err != nil {
+			log.Errorf("failed during fetching the restore object, %v", err)
```
Comment (lsrinivas-pure) [inline drivers/pds/pdsrestore/restore_utils.go:69]: do not log the err if you are returning. as this be logged in calling function also

File: drivers/pds/pdsrestore/restore_utils.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+			return false, err
+		}
+		log.Infof("Restore status -  %v", state)
+		if strings.ToLower(state) != strings.ToLower("Successful") {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		log.Errorf("restore failed {polling}, %v", err)
```
Comment (lsrinivas-pure) [inline drivers/pds/pdsrestore/restore_utils.go:79]: do not log the err if you are returning. as this be logged in calling function also

File: drivers/pds/pdsrestore/restore_utils.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+	bkpResourceConfig := resourceStructToMap(bkpDsEntity.Deployment.Resources)
+	restoreResourceConfig := resourceStructToMap(restoreDsEntity.Deployment.Resources)
+	log.Infof("Backed up resource configuration- %v", bkpResourceConfig)
+	log.Infof("Restored resource configuration- %v", restoreResourceConfig)
+	if !compareMaps(bkpResourceConfig, restoreResourceConfig) {
+		return fmt.Errorf("restored resource configuration are not same as backed up resource config")
+	}
+
+	// Validate the StorageOption configuration
+	log.Info("Validating StorageOption config post restore.")
```
Comment (lsrinivas-pure) [inline drivers/pds/pdsrestore/restore_utils.go:138]: add apace after storage

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

**Review by aravindranath14 — APPROVED** [review]
LGTM

File: drivers/pds/pdsrestore/restore_utils.go (hunk 7 above)
Comment (gauraa2) [reply drivers/pds/pdsrestore/restore_utils.go]: As part of the next PR, will update error logging for all.

**Comment by gauraa2** [comment]
As part of next PR, will address the error logging covering wider scope.
cc: @lsrinivas-pure

**Review by lsrinivas-pure — APPROVED** [review]

**Pull request merged**